packages/fxa-auth-server/lib/payments/stripe.ts

```diff
@@ -678,6 +679,127 @@ export class StripeHelper {
     }
 
     // Is the coupon valid for this price?
+    const planContainsPromo = await this.checkPromotionCodeForPlan(
+      code,
+      priceId
+    );
+    if (!planContainsPromo) {
+      return;
+    }
+
+    return promotionCode;
+  }
+
+  /**
+   * Retrieve details about a coupon for a given priceId and possible
+   * promotion code for a customer in the provided country. Will also
+   * provide the discount amount for the subscription via
+   * previewInvoice return value. Coupon details are returned
+   * regardless of current validity (expiry, redeemability).
+   *
+   * Throws invalidPromoCode error if the promotion code does not
+   * exist for the provided priceId.
+   */
+  async retrieveCouponDetails({
+    country,
+    priceId,
+    promotionCode,
+  }: {
+    country: string;
+    priceId: string;
+    promotionCode: string;
+  }): Promise<Coupon.couponDetailsSchema> {
+    try {
+      const stripePromotionCode = await this.retrievePromotionCodeForPlan(
+        promotionCode,
+        priceId
+      );
+
+      if (stripePromotionCode?.coupon.id) {
+        let invoice: Stripe.Invoice | undefined = undefined;
```

Comment:
it needs to be let because it can be undefined and we dont do the assignment until later on. theres actually a few other places in this file where we employ the same pattern (see the payInvoice method). we basically want to assume its undefined, but if we cant retrieve it, thats actually ok which is why we swallow the error.

Reply:
we could refactor the previewInvoice method to not throw an error if it isnt possible to retrieve the invoice with the provided coupon data. the problem arises from the fact that we are trying to retrieve a coupon that may be invalid but we want to be able to tell the FE why it is invalid.

Reply:
Ah OK I see what you mean. 

IMO, as best practice I think it'd be best to use `const` where possible. Some of the benefits are discussed [here](https://charles-stover.medium.com/replacing-let-with-const-86797b790775).

I left some comments with code recommendations in another thread that show how we could accomplish that.

Reply:
see other comments - no longer using let